File: internal/status/status.go
// Package status assembles the read-only workspace status snapshot emitted by
// `qualitymd status`.
package status

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"slices"

	"github.com/qualitymd/quality.md/internal/evaluation"
	"github.com/qualitymd/quality.md/internal/lint"
	"github.com/qualitymd/quality.md/internal/model"
	"github.com/qualitymd/quality.md/internal/receipt"
	"github.com/qualitymd/quality.md/internal/workspace"
)

// SchemaVersion is the current status snapshot schema version.
const SchemaVersion = 2

// Readiness identifies the current model/evaluation lifecycle state.
type Readiness string

const (
	ReadinessMissingModel             Readiness = "missing-model"
	ReadinessInvalidModel             Readiness = "invalid-model"
	ReadinessReadyToEvaluate          Readiness = "ready-to-evaluate"
	ReadinessHasEvaluationHistory     Readiness = "has-evaluation-history"
	ReadinessNeedsEvaluationReconcile Readiness = "needs-evaluation-reconciliation"
)

// Options configures a status snapshot.
type Options struct {
	Path string
}

// WorkspaceSnapshot is the JSON contract emitted by qualitymd status.
type WorkspaceSnapshot struct {
	SchemaVersion int               `json:"schemaVersion"`
	Path          string            `json:"path"`
	Workspace     *WorkspaceStatus  `json:"workspace,omitempty"`
	Readiness     Readiness         `json:"readiness"`
	Model         ModelStatus       `json:"model"`
	Evaluations   EvaluationHistory `json:"evaluations"`
	NextActions   []receipt.Action  `json:"nextActions"`
}

// WorkspaceStatus summarizes the resolved workspace paths for the selected
// model. All paths are repository-relative or workspace-relative, never
// absolute host paths.
type WorkspaceStatus struct {
	Root          string `json:"root"`
	Model         string `json:"model"`
	Config        string `json:"config"`
	ConfigPresent bool   `json:"configPresent"`
	DataDir       string `json:"dataDir"`
	EvaluationDir string `json:"evaluationDir"`
	ChangelogDir  string `json:"changelogDir"`
	LogDir        string `json:"logDir"`
}

// ModelStatus summarizes whether the model is present, valid, and evaluable.
type ModelStatus struct {
	Present        bool             `json:"present"`
	Valid          bool             `json:"valid"`
	Lint           LintStatus       `json:"lint"`
	Shape          *ModelShape      `json:"shape,omitempty"`
	SourceCoverage []SourceCoverage `json:"sourceCoverage,omitempty"`
}

// LintStatus embeds lint summary and findings in status output.
type LintStatus struct {
	Summary  lint.Summary   `json:"summary"`
	Findings []lint.Finding `json:"findings,omitempty"`
}

// ModelShape counts the model's structural elements.
type ModelShape struct {
	Areas        int `json:"areas"`
	Factors      int `json:"factors"`
	Requirements int `json:"requirements"`
	RatingLevels int `json:"ratingLevels"`
}

// SourceCoverage summarizes source declarations for one area. The source
// state comes from the shared model resolver (model.EffectiveSource).
type SourceCoverage struct {
	AreaPath     []string          `json:"areaPath"`
	Label        string            `json:"label"`
	SourceState  model.SourceState `json:"sourceState"`
	Source       string            `json:"source,omitempty"`
	Factors      int               `json:"factors"`
	Requirements int               `json:"requirements"`
	ChildAreas   int               `json:"childAreas"`
}

// EvaluationHistory summarizes discovered evaluation runs for the model.
type EvaluationHistory struct {
	Path    string                 `json:"path,omitempty"`
	Runs    int                    `json:"runs"`
	Latest  *EvaluationRunSummary  `json:"latest,omitempty"`
	Items   []EvaluationRunSummary `json:"items"`
	Summary EvaluationSummary      `json:"summary"`
}

// EvaluationSummary aggregates evaluation history counts.
type EvaluationSummary struct {
	Reportable int `json:"reportable"`
	Incomplete int `json:"incomplete"`
	// AwaitingEvaluator counts incomplete runs checkpointed at a pending
	// harness work request. They are included in Incomplete as well.
	AwaitingEvaluator int `json:"awaitingEvaluator,omitempty"`
	Stale             int `json:"stale"`
	Problems          int `json:"problems"`
}

// EvaluationRunSummary summarizes one evaluation run in status output.
type EvaluationRunSummary struct {
	Path       string `json:"path"`
	Reportable bool   `json:"reportable"`
	// Lifecycle is the runner lifecycle status of an artifact-backed run;
	// empty for manual multi-file runs. awaiting_evaluator means the run is
	// resumable with harness judgment as the pending action, not failed or
	// generically incomplete.
	Lifecycle     string `json:"lifecycle,omitempty"`
	Stale         bool   `json:"stale"`
	DataArtifacts int    `json:"dataArtifacts"`
	Gaps          int    `json:"gaps"`
	Problem       string `json:"problem,omitempty"`
}

// Snapshot assembles a deterministic workspace status snapshot.
func Snapshot(opts Options) (*WorkspaceSnapshot, error) {
	path := opts.Path
	if path == "" {
		path = "QUALITY.md"
	}
	result := &WorkspaceSnapshot{
		SchemaVersion: SchemaVersion,
		Path:          path,
		Evaluations: EvaluationHistory{
			Items: []EvaluationRunSummary{},
		},
	}

	modelBytes, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			result.Readiness = ReadinessMissingModel
			result.Model = ModelStatus{Present: false}
			result.NextActions = []receipt.Action{{
				ID:      "init",
				Label:   "Create a starter QUALITY.md",
				Command: "qualitymd init " + path,
			}}
			return result, nil
		}
		return nil, err
	}

	lintResult, err := lint.Check(path)
	if err != nil {
		return nil, err
	}
	result.Model = ModelStatus{
		Present: true,
		Valid:   lintResult.Valid,
		Lint: LintStatus{
			Summary:  lintResult.Summary,
			Findings: lintResult.Findings,
		},
	}
	if !lintResult.Valid {
		result.Readiness = ReadinessInvalidModel
		result.NextActions = invalidModelActions(path, lintResult.Summary)
		return result, nil
	}

	spec, err := lint.Load(path)
	if err != nil {
		return nil, err
	}
	shape, coverage := modelSnapshot(spec)
	result.Model.Shape = &shape
	result.Model.SourceCoverage = coverage

	ws, err := workspace.Resolve(workspace.Options{
		Model: path,
	})
	if err != nil {
		return nil, err
	}
	result.Workspace = workspaceStatus(ws)

	history, err := evaluationHistory(ws, modelBytes)
	if err != nil {
		return nil, err
	}
	result.Evaluations = history
	result.Readiness = readiness(history)
	result.NextActions = nextActions(path, history, result.Readiness)
	return result, nil
}

func workspaceStatus(ws *workspace.Workspace) *WorkspaceStatus {
	return &WorkspaceStatus{
		Root:          ws.WorkspaceRoot.RepoRel,
		Model:         ws.Model.Rel,
		Config:        ws.Config.Rel,
		ConfigPresent: ws.ConfigPresent,
		DataDir:       ws.DataDir.Rel,
		EvaluationDir: ws.Evaluations.Rel,
		ChangelogDir:  ws.Log.Rel,
		LogDir:        ws.FeedbackLog.Rel,
	}
}

func invalidModelActions(path string, summary lint.Summary) []receipt.Action {
	if summary.Fixable > 0 {
		return []receipt.Action{{
			ID:      "fix",
			Label:   "Apply deterministic lint repairs",
			Command: "qualitymd lint --fix " + path,
		}}
	}
	return []receipt.Action{{
		ID:      "lint",
		Label:   "Review lint findings",
		Command: "qualitymd lint " + path,
	}}
}

func modelSnapshot(spec *model.Spec) (ModelShape, []SourceCoverage) {
	return modelShape(spec), sourceCoverage(spec)
}

// modelShape counts the model's structural elements from the shared projection,
// so status and the model command read one model-tree walk.
func modelShape(spec *model.Spec) ModelShape {
	shape := ModelShape{RatingLevels: len(spec.RatingScale)}
	for _, element := range model.Flatten(model.Project(spec)) {
		switch element.Kind {
		case model.KindArea:
			shape.Areas++
		case model.KindFactor:
			shape.Factors++
		case model.KindRequirement:
			shape.Requirements++
		}
	}
	return shape
}

// sourceCoverage builds the per-area source provenance rows. Source state is a
// status-only concern the identity projection deliberately omits, so it keeps
// its own source-aware walk over the shared model resolver.
func sourceCoverage(spec *model.Spec) []SourceCoverage {
	label := spec.Title
	if label == "" {
		label = "Model"
	}
	rows := []SourceCoverage{sourceCoverageRow(spec, nil, label, len(spec.Factors), len(spec.Requirements), len(spec.Areas))}
	return appendAreaCoverage(rows, spec, spec.Areas, nil)
}

func appendAreaCoverage(rows []SourceCoverage, spec *model.Spec, areas map[string]model.Area, parentPath []string) []SourceCoverage {
	for _, name := range sortedKeys(areas) {
		area := areas[name]
		path := appendString(parentPath, name)
		label := area.Title
		if label == "" {
			label = name
		}
		rows = append(rows, sourceCoverageRow(spec, path, label, len(area.Factors), len(area.Requirements), len(area.Areas)))
		rows = appendAreaCoverage(rows, spec, area.Areas, path)
	}
	return rows
}

func sourceCoverageRow(spec *model.Spec, path []string, label string, factors, requirements, children int) SourceCoverage {
	row := SourceCoverage{
		AreaPath:     append([]string{}, path...),
		Label:        label,
		Factors:      factors,
		Requirements: requirements,
		ChildAreas:   children,
	}
	selector, state := model.EffectiveSource(spec, model.AreaPath(path))
	row.SourceState = state
	if state != model.SourceStateDefault {
		// The default selector is implicit (the document's directory); the row
		// reports it through the state alone.
		row.Source = selector
	}
	return row
}

func sortedKeys[V any](m map[string]V) []string {
	keys := make([]string, 0, len(m))
	for key := range m {
		keys = append(keys, key)
	}
	slices.Sort(keys)
	return keys
}

func appendString(path []string, value string) []string {
	out := make([]string, 0, len(path)+1)
	out = append(out, path...)
	out = append(out, value)
	return out
}

func evaluationHistory(ws *workspace.Workspace, modelBytes []byte) (EvaluationHistory, error) {
	history := EvaluationHistory{Items: []EvaluationRunSummary{}}
	evalDirAbs := ws.Evaluations.Abs
	evalDirRel := ws.Evaluations.Rel
	history.Path = evalDirRel
	runs, err := evaluation.ListRunDirs(evalDirAbs, evalDirRel)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return history, nil
		}
		return history, err
	}
	for _, runDir := range runs {
		summary := inspectRun(runDir, modelBytes)
		history.Items = append(history.Items, summary)
		history.Runs++
		if summary.Problem != "" {
			history.Summary.Problems++
		}
		if summary.Reportable {
			history.Summary.Reportable++
		} else {
			history.Summary.Incomplete++
			if summary.Lifecycle == evaluation.RunStatusAwaitingEvaluator {
				history.Summary.AwaitingEvaluator++
			}
		}
		if summary.Stale {
			history.Summary.Stale++
		}
	}
	if len(history.Items) > 0 {
		latest := history.Items[len(history.Items)-1]
		history.Latest = &latest
	}
	return history, nil
}

func inspectRun(runDir evaluation.RunDir, modelBytes []byte) EvaluationRunSummary {
	summary := EvaluationRunSummary{Path: runDir.Rel}
	runModel, err := os.ReadFile(filepath.Join(runDir.Abs, evaluation.ModelSnapshotFile))
	if err != nil {
		summary.Problem = fmt.Sprintf("reading %s: %v", evaluation.ModelSnapshotFile, err)
		return summary
	}
	summary.Stale = string(runModel) != string(modelBytes)

	run, err := evaluation.Inspect(runDir.Abs)
	if err != nil {
		summary.Problem = err.Error()
		return summary
	}
	runStatus := run.Status()
	summary.Reportable = runStatus.Reportable
	summary.Lifecycle = runStatus.Lifecycle
	summary.DataArtifacts = runStatus.Data.Artifacts
	summary.Gaps = len(runStatus.Gaps)
	return summary
}

func readiness(history EvaluationHistory) Readiness {
	if history.Runs == 0 {
		return ReadinessReadyToEvaluate
	}
	if history.Summary.Incomplete > 0 || history.Summary.Stale > 0 || history.Summary.Problems > 0 {
		return ReadinessNeedsEvaluationReconcile
	}
	return ReadinessHasEvaluationHistory
}

func nextActions(path string, history EvaluationHistory, state Readiness) []receipt.Action {
	switch state {
	case ReadinessReadyToEvaluate:
		return []receipt.Action{{
			ID:      "evaluation-create",
			Label:   "Create an evaluation run",
			Command: "qualitymd evaluation create --model " + path,
		}}
	case ReadinessNeedsEvaluationReconcile:
		return reconciliationActions(path, history)
	case ReadinessHasEvaluationHistory:
		if history.Latest != nil && history.Latest.Reportable {
			return []receipt.Action{{
				ID:      "report-build",
				Label:   "Build the latest evaluation report",
				Command: "qualitymd evaluation report build --model " + path + " " + history.Latest.Path,
			}}
		}
	}
	return []receipt.Action{}
}

func reconciliationActions(path string, history EvaluationHistory) []receipt.Action {
	if history.Latest == nil {
		return nil
	}
	latest := history.Latest
	switch {
	case latest.Problem == "" && latest.Lifecycle == evaluation.RunStatusAwaitingEvaluator:
		// Awaiting harness judgment is a normal checkpoint: the continuation
		// is resuming the run to recover the pending work request.
		return []receipt.Action{{
			ID:      "evaluation-run-reemit",
			Label:   "Resume the awaiting evaluation run to recover its pending work request",
			Command: "qualitymd evaluation run --model " + path + " --resume " + latest.Path + " --json",
		}}
	case latest.Problem != "" || !latest.Reportable:
		return []receipt.Action{{
			ID:      "evaluation-status-latest",
			Label:   "Inspect the latest evaluation run",
			Command: "qualitymd evaluation status --model " + path + " " + latest.Path,
		}}
	case latest.Stale:
		return []receipt.Action{{
			ID:      "evaluation-create",
			Label:   "Create a fresh evaluation run",
			Command: "qualitymd evaluation create --model " + path,
		}}
	default:
		return nil
	}
}

File: src/application/status.ts
import * as Effect from "effect/Effect"
import * as FileSystem from "effect/FileSystem"
import * as Path from "effect/Path"

import { FileSystemFailure } from "../domain/errors.ts"
import { commandResult, ExitCode, type CommandResult } from "../domain/command-result.ts"
import { jsonDocument } from "../domain/json.ts"
import { invalidDocumentResult, lintDocument } from "../domain/lint/lint.ts"
import type { Action, LintResult } from "../domain/lint/result.ts"
import { parseQualityDocument, QualityDocumentParseError } from "../domain/model/document.ts"
import {
  decodeModel,
  effectiveSource,
  flattenModel,
  projectModel,
  type Area,
  type QualityModel,
} from "../domain/model/model.ts"
import { resolveWorkspace, type Workspace } from "../services/workspace.ts"

type Readiness =
  | "missing-model"
  | "invalid-model"
  | "ready-to-evaluate"
  | "has-evaluation-history"
  | "needs-evaluation-reconciliation"

interface EvaluationRunSummary {
  readonly path: string
  readonly reportable: boolean
  readonly lifecycle?: string
  readonly stale: boolean
  readonly dataArtifacts: number
  readonly gaps: number
  readonly problem?: string
}

interface EvaluationHistory {
  readonly path?: string
  readonly runs: number
  readonly latest?: EvaluationRunSummary
  readonly items: ReadonlyArray<EvaluationRunSummary>
  readonly summary: {
    readonly reportable: number
    readonly incomplete: number
    readonly awaitingEvaluator?: number
    readonly stale: number
    readonly problems: number
  }
}

interface StatusSnapshot {
  readonly schemaVersion: 2
  readonly path: string
  readonly workspace?: {
    readonly root: string
    readonly model: string
    readonly config: string
    readonly configPresent: boolean
    readonly dataDir: string
    readonly evaluationDir: string
    readonly changelogDir: string
    readonly logDir: string
  }
  readonly readiness: Readiness
  readonly model: Record<string, unknown>
  readonly evaluations: EvaluationHistory
  readonly nextActions: ReadonlyArray<Action>
}

const emptyHistory = (path?: string): EvaluationHistory => ({
  ...(path === undefined ? {} : { path }),
  runs: 0,
  items: [],
  summary: { reportable: 0, incomplete: 0, stale: 0, problems: 0 },
})

const parseLint = (
  path: string,
  raw: string,
): { readonly result: LintResult; readonly model?: QualityModel } => {
  try {
    const document = parseQualityDocument(path, raw)
    const result = lintDocument(document).result
    return { result, ...(result.valid ? { model: decodeModel(document) } : {}) }
  } catch (cause) {
    if (cause instanceof QualityDocumentParseError) return { result: invalidDocumentResult(path) }
    throw cause
  }
}

const sourceCoverage = (model: QualityModel) => {
  const rows: Array<Record<string, unknown>> = []
  const visit = (area: Area, path: ReadonlyArray<string>, fallback: string) => {
    const resolved = effectiveSource(model, path)
    rows.push({
      areaPath: path,
      label: area.title || fallback,
      sourceState: resolved.state,
      ...(resolved.state === "default" ? {} : { source: resolved.selector }),
      factors: Object.keys(area.factors ?? {}).length,
      requirements: Object.keys(area.requirements ?? {}).length,
      childAreas: Object.keys(area.areas ?? {}).length,
    })
    for (const [name, child] of Object.entries(area.areas ?? {}).sort(([a], [b]) =>
      a.localeCompare(b),
    )) {
      visit(child, [...path, name], name)
    }
  }
  visit(model, [], "Model")
  return rows
}

const modelShape = (model: QualityModel) => {
  const shape = { areas: 0, factors: 0, requirements: 0, ratingLevels: model.ratingScale.length }
  for (const element of flattenModel(projectModel(model))) {
    if (element.kind === "area") shape.areas += 1
    else if (element.kind === "factor") shape.factors += 1
    else shape.requirements += 1
  }
  return shape
}

const inspectArtifact = (
  path: string,
  currentModel: string,
  snapshot: string,
  artifactRaw: string,
): EvaluationRunSummary => {
  try {
    const artifact = JSON.parse(artifactRaw) as {
      readonly state?: {
        readonly status?: string
        readonly workUnits?: Readonly<Record<string, { readonly status?: string }>>
        readonly pendingEvaluatorCalls?: ReadonlyArray<unknown>
      }
      readonly results?: { readonly payloads?: ReadonlyArray<unknown> }
    }
    const lifecycle = artifact.state?.status ?? ""
    const gaps = Object.values(artifact.state?.workUnits ?? {}).filter(
      (unit) => unit.status !== "completed",
    ).length
    return {
      path,
      reportable: gaps === 0 && lifecycle === "completed",
      ...(lifecycle === "" ? {} : { lifecycle }),
      stale: snapshot !== currentModel,
      dataArtifacts: artifact.results?.payloads?.length ?? 0,
      gaps,
    }
  } catch (cause) {
    return {
      path,
      reportable: false,
      stale: snapshot !== currentModel,
      dataArtifacts: 0,
      gaps: 0,
      problem: cause instanceof Error ? cause.message : String(cause),
    }
  }
}

const evaluationHistory = (workspace: Workspace, currentModel: string) =>
  Effect.gen(function* () {
    const fs = yield* FileSystem.FileSystem
    const paths = yield* Path.Path
    if (!(yield* fs.exists(workspace.evaluations.abs)))
      return emptyHistory(workspace.evaluations.rel)
    const runs: Array<{ readonly number: number; readonly name: string }> = []
    for (const name of yield* fs.readDirectory(workspace.evaluations.abs)) {
      const runAbs = paths.join(workspace.evaluations.abs, name)
      if ((yield* fs.stat(runAbs)).type !== "Directory") continue
      const artifactPath = paths.join(runAbs, "evaluation.json")
      if (yield* fs.exists(artifactPath)) {
        try {
          const artifact = JSON.parse(yield* fs.readFileString(artifactPath)) as {
            readonly manifest?: { readonly run?: { readonly number?: number } }
          }
          runs.push({ number: artifact.manifest?.run?.number ?? Number(name.slice(0, 4)), name })
        } catch {
          runs.push({ number: Number(name.slice(0, 4)), name })
        }
        continue
      }
      const match = /^(\d{4})-([a-z0-9-]+)-eval$/.exec(name)
      if (match === null || match[2]!.split("-").includes("quality")) continue
      runs.push({ number: Number(match[1]), name })
    }
    runs.sort((a, b) => a.number - b.number || a.name.localeCompare(b.name))
    const items: Array<EvaluationRunSummary> = []
    for (const run of runs) {
      const runAbs = paths.join(workspace.evaluations.abs, run.name)
      const path = `${workspace.evaluations.rel}/${run.name}`
      try {
        const snapshot = yield* fs.readFileString(paths.join(runAbs, "model-snapshot.md"))
        const artifactPath = paths.join(runAbs, "evaluation.json")
        if (yield* fs.exists(artifactPath)) {
          items.push(
            inspectArtifact(path, currentModel, snapshot, yield* fs.readFileString(artifactPath)),
          )
        } else {
          const data = paths.join(runAbs, "data")
          const artifacts = (yield* fs.exists(data))
            ? (yield* fs.readDirectory(data)).filter((name) => name.endsWith(".json")).length
            : 0
          items.push({
            path,
            reportable: false,
            stale: snapshot !== currentModel,
            dataArtifacts: artifacts,
            gaps: 1,
          })
        }
      } catch (cause) {
        items.push({
          path,
          reportable: false,
          stale: false,
          dataArtifacts: 0,
          gaps: 0,
          problem: cause instanceof Error ? cause.message : String(cause),
        })
      }
    }
    const summary = {
      reportable: items.filter((item) => item.reportable).length,
      incomplete: items.filter((item) => !item.reportable).length,
      awaitingEvaluator: items.filter((item) => item.lifecycle === "awaiting_evaluator").length,
      stale: items.filter((item) => item.stale).length,
      problems: items.filter((item) => item.problem !== undefined).length,
    }
    return {
      path: workspace.evaluations.rel,
      runs: items.length,
      ...(items.length === 0 ? {} : { latest: items.at(-1)! }),
      items,
      summary: {
        reportable: summary.reportable,
        incomplete: summary.incomplete,
        ...(summary.awaitingEvaluator === 0
          ? {}
          : { awaitingEvaluator: summary.awaitingEvaluator }),
        stale: summary.stale,
        problems: summary.problems,
      },
    } satisfies EvaluationHistory
  })

const actionsFor = (
  path: string,
  readiness: Readiness,
  history: EvaluationHistory,
): ReadonlyArray<Action> => {
  if (readiness === "ready-to-evaluate") {
    return [
      {
        id: "evaluation-create",
        label: "Create an evaluation run",
        command: `qualitymd evaluation create --model ${path}`,
      },
    ]
  }
  const latest = history.latest
  if (readiness === "needs-evaluation-reconciliation" && latest !== undefined) {
    if (latest.problem === undefined && latest.lifecycle === "awaiting_evaluator") {
      return [
        {
          id: "evaluation-run-reemit",
          label: "Resume the awaiting evaluation run to recover its pending work request",
          command: `qualitymd evaluation run --model ${path} --resume ${latest.path} --json`,
        },
      ]
    }
    if (latest.problem !== undefined || !latest.reportable) {
      return [
        {
          id: "evaluation-status-latest",
          label: "Inspect the latest evaluation run",
          command: `qualitymd evaluation status --model ${path} ${latest.path}`,
        },
      ]
    }
    if (latest.stale) {
      return [
        {
          id: "evaluation-create",
          label: "Create a fresh evaluation run",
          command: `qualitymd evaluation create --model ${path}`,
        },
      ]
    }
  }
  if (readiness === "has-evaluation-history" && latest?.reportable === true) {
    return [
      {
        id: "report-build",
        label: "Build the latest evaluation report",
        command: `qualitymd evaluation report build --model ${path} ${latest.path}`,
      },
    ]
  }
  return []
}

const snapshot = (path: string) =>
  Effect.gen(function* () {
    const fs = yield* FileSystem.FileSystem
    if (!(yield* fs.exists(path))) {
      return {
        schemaVersion: 2,
        path,
        readiness: "missing-model",
        model: { present: false },
        evaluations: emptyHistory(),
        nextActions: [
          { id: "init", label: "Create a starter QUALITY.md", command: `qualitymd init ${path}` },
        ],
      } satisfies StatusSnapshot
    }
    const raw = yield* fs.readFileString(path)
    const parsed = parseLint(path, raw)
    const lint = {
      summary: parsed.result.summary,
      ...(parsed.result.findings.length === 0 ? {} : { findings: parsed.result.findings }),
    }
    if (!parsed.result.valid || parsed.model === undefined) {
      return {
        schemaVersion: 2,
        path,
        readiness: "invalid-model",
        model: { present: true, valid: false, lint },
        evaluations: emptyHistory(),
        nextActions: [
          parsed.result.summary.fixable > 0
            ? {
                id: "fix",
                label: "Apply deterministic lint repairs",
                command: `qualitymd lint --fix ${path}`,
              }
            : { id: "lint", label: "Review lint findings", command: `qualitymd lint ${path}` },
        ],
      } satisfies StatusSnapshot
    }
    const workspace = yield* resolveWorkspace({ model: path })
    const history = yield* evaluationHistory(workspace, raw)
    const readiness: Readiness =
      history.runs === 0
        ? "ready-to-evaluate"
        : history.summary.incomplete > 0 ||
            history.summary.stale > 0 ||
            history.summary.problems > 0
          ? "needs-evaluation-reconciliation"
          : "has-evaluation-history"
    return {
      schemaVersion: 2,
      path,
      workspace: {
        root: workspace.workspaceRoot.repoRel,
        model: workspace.model.rel,
        config: workspace.config.rel,
        configPresent: workspace.configPresent,
        dataDir: workspace.dataDir.rel,
        evaluationDir: workspace.evaluations.rel,
        changelogDir: workspace.changelog.rel,
        logDir: workspace.logs.rel,
      },
      readiness,
      model: {
        present: true,
        valid: true,
        lint,
        shape: modelShape(parsed.model),
        sourceCoverage: sourceCoverage(parsed.model),
      },
      evaluations: history,
      nextActions: actionsFor(path, readiness, history),
    } satisfies StatusSnapshot
  }).pipe(
    Effect.mapError((cause) =>
      cause instanceof FileSystemFailure
        ? cause
        : new FileSystemFailure({ detail: cause instanceof Error ? cause.message : String(cause) }),
    ),
  )

const renderHuman = (value: StatusSnapshot) => {
  const lint = value.model.lint as { readonly summary: LintResult["summary"] } | undefined
  const present = value.model.present === true
  const valid = value.model.valid === true
  const model = !present
    ? "absent"
    : valid
      ? "present, valid"
      : `present, invalid (${lint?.summary.errors ?? 0} error(s), ${lint?.summary.warnings ?? 0} warning(s))`
  const shape = value.model.shape as
    | { readonly areas: number; readonly factors: number; readonly requirements: number }
    | undefined
  let output = "Workspace Status\n"
  if (value.workspace !== undefined) output += `- Workspace: ${value.workspace.root}\n`
  output += `- Model file: ${value.workspace?.model ?? value.path}: ${model}\n`
  if (shape !== undefined)
    output += `- Model: ${shape.areas} area(s), ${shape.factors} factor(s), ${shape.requirements} requirement(s)\n`
  output += `- Evaluation history: ${value.evaluations.runs} run(s), ${value.evaluations.summary.incomplete} incomplete, ${value.evaluations.summary.stale} stale\n`
  output += `- Readiness: ${value.readiness}\n`
  return output
}

export const statusCommand = (input: {
  readonly path: string
  readonly json: boolean
}): Effect.Effect<CommandResult, FileSystemFailure, FileSystem.FileSystem | Path.Path> => {
  if (input.path === "-") {
    return Effect.succeed(
      commandResult("", {
        stderr: "qualitymd: status does not read from stdin; pass a file path\n",
        exitCode: ExitCode.usage,
      }),
    )
  }
  return snapshot(input.path).pipe(
    Effect.map((value) => {
      const action = value.nextActions.at(0)
      return commandResult(input.json ? jsonDocument(value) : renderHuman(value), {
        stderr: input.json || action === undefined ? "" : `\nNext: ${action.command}\n`,
      })
    }),
  )
}
